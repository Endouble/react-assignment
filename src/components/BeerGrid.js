import React from 'react';
import PropTypes from 'prop-types';
import { Grid, Card, Image, Message } from 'semantic-ui-react';
import BeerDetails from './BeerDetails';
import '../styles/BeerGrid.css';

class BeerGrid extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            isModalOpened: false,
        };

        this.beerDetailData = null;
        this.onCloseDetail = this.onCloseDetail.bind(this);
    }

    onCloseDetail() {
        this.beerDetailData = null;
        this.setState({
            isModalOpened: false,
        });
    }

    handleClick(event, beer) {
        event.preventDefault();

        this.openBeerDetail(beer);
    }

    handleKeyPress(event, beer) {
        event.preventDefault();

        if (event.key === 'Enter') {
            this.openBeerDetail(beer);
        }
    }

    openBeerDetail(beer) {
        this.beerDetailData = beer;
        this.setState({
            isModalOpened: true,
        });
    }

    render() {
        const { beers } = this.props;
        const { isModalOpened } = this.state;

        return (
            <React.Fragment>
                {
                    beers.length > 0
                        ? (
                            <Grid stackable columns={4}>
                                {
                                    beers.map(beer => (
                                        <Grid.Column key={beer.id} stretched>
                                            <Card centered link onClick={e => this.handleClick(e, beer)} onKeyPress={e => this.handleKeyPress(e, beer)} tabIndex='0'>
                                                <div className="card-image-wrapper">
                                                    <Image
                                                        centered
                                                        style={{ maxHeight: 200 }}
                                                        src={beer.image_url}
                                                        alt={beer.name}
                                                    />
                                                </div>
                                                <Card.Content>
                                                    <Card.Header>{beer.name}</Card.Header>
                                                    <Card.Meta>
                                                        <span className='date'>{`Since ${beer.first_brewed}`}</span>
                                                    </Card.Meta>
                                                    <Card.Description>{beer.tagline}</Card.Description>
                                                </Card.Content>
                                                <Card.Content extra>
                                                    { `${beer.abv} % - ${beer.ibu} IBU` }
                                                </Card.Content>
                                            </Card>
                                        </Grid.Column>
                                    ))
                                }
                            </Grid>
                        )
                        : <Message color='orange'>{'We didn\'t find beers that match your taste!'}</Message>
                }
                {
                    isModalOpened
                    && <BeerDetails beer={this.beerDetailData} onCloseDetail={this.onCloseDetail} />
                }
            </React.Fragment>
        );
    }
}

BeerGrid.defaultProps = {
    beers: [],
};

BeerGrid.propTypes = {
    beers: PropTypes.arrayOf(PropTypes.shape({})),
};

export default BeerGrid;
