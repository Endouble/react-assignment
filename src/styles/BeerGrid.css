.card-image-wrapper {
    padding: 14px 0;
}
