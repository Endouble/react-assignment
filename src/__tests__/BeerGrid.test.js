import React from 'react';
import { shallow } from 'enzyme';
import BeerGrid from '../components/BeerGrid';

describe('BeerGrid Unit Test', () => {
    let props;
    let shallowedBeerGrid;

    const getBeerGrid = () => {
        if (!shallowedBeerGrid) {
            shallowedBeerGrid = shallow(<BeerGrid {...props} />);
        }
        return shallowedBeerGrid;
    };

    beforeEach(() => {
        props = {
            beers: [
                {
                    id: 1,
                    name: 'Test Beer',
                    tagline: 'A beer for testing',
                    description: 'Just a mock beer for testing',
                    first_brewed: 'Just a mock beer for testing',
                    image_url: 'https://testurl.com/beer/image.png',
                    food_pairing: ['Every snack you can imagine'],
                    brewers_tips: 'Drink carefully',
                    abv: 5.0,
                    ibu: 40,
                },
            ],
        };
        shallowedBeerGrid = undefined;
    });

    test('Component should render without crashing', () => {
        expect(getBeerGrid()).toBeDefined();
    });

    test('Component should render correctly', () => {
        expect(getBeerGrid()).toMatchSnapshot();
    });

    test('Component should render message when beers is empty', () => {
        props.beers = [];
        const wrapper = getBeerGrid();
        expect(wrapper.find('Message')).toBeDefined();
    });

    test('Component should render beers cards', () => {
        const wrapper = getBeerGrid();
        expect(wrapper.find('Card')).toHaveLength(props.beers.length);
    });

    test('Component should render beer details on click in a beer card', () => {
        const wrapper = getBeerGrid();
        const firstCard = wrapper.find('Card').at(0);
        firstCard.simulate('click', { preventDefault: jest.fn() });
        expect(wrapper.find('BeerDetails')).toBeDefined();
        wrapper.instance().onCloseDetail();
        expect(wrapper.find('BeerDetails')).toHaveLength(0);
    });

    test('Component should render beer details on keypress Enter in a beer card', () => {
        const wrapper = getBeerGrid();
        const firstCard = wrapper.find('Card').at(0);
        firstCard.simulate('focus');
        firstCard.simulate('keypress', { preventDefault: jest.fn(), key: 'Enter' });
        expect(wrapper.find('BeerDetails')).toBeDefined();
        wrapper.instance().onCloseDetail();
        expect(wrapper.find('BeerDetails')).toHaveLength(0);
    });

    test('Component should not call openBeerDetail function on keypress different from Enter', () => {
        const wrapper = getBeerGrid();
        wrapper.instance().openBeerDetail = jest.fn();
        const firstCard = wrapper.find('Card').at(0);
        firstCard.simulate('focus');
        firstCard.simulate('keypress', { preventDefault: jest.fn(), key: 'a' });
        expect(wrapper.instance().openBeerDetail).not.toHaveBeenCalled();
    });
});
