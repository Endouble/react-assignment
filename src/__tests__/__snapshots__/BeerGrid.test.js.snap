// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BeerGrid Unit Test Component should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <BeerGrid
    beers={
      Array [
        Object {
          "abv": 5,
          "brewers_tips": "Drink carefully",
          "description": "Just a mock beer for testing",
          "first_brewed": "Just a mock beer for testing",
          "food_pairing": Array [
            "Every snack you can imagine",
          ],
          "ibu": 40,
          "id": 1,
          "image_url": "https://testurl.com/beer/image.png",
          "name": "Test Beer",
          "tagline": "A beer for testing",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Grid
          columns={4}
          stackable={true}
        >
          <GridColumn
            stretched={true}
          >
            <Card
              centered={true}
              link={true}
              onClick={[Function]}
              onKeyPress={[Function]}
              tabIndex="0"
            >
              <div
                className="card-image-wrapper"
              >
                <Image
                  alt="Test Beer"
                  as="img"
                  centered={true}
                  src="https://testurl.com/beer/image.png"
                  style={
                    Object {
                      "maxHeight": 200,
                    }
                  }
                  ui={true}
                />
              </div>
              <CardContent>
                <CardHeader>
                  Test Beer
                </CardHeader>
                <CardMeta>
                  <span
                    className="date"
                  >
                    Since Just a mock beer for testing
                  </span>
                </CardMeta>
                <CardDescription>
                  A beer for testing
                </CardDescription>
              </CardContent>
              <CardContent
                extra={true}
              >
                5 % - 40 IBU
              </CardContent>
            </Card>
          </GridColumn>
        </Grid>,
        false,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <GridColumn
              stretched={true}
            >
              <Card
                centered={true}
                link={true}
                onClick={[Function]}
                onKeyPress={[Function]}
                tabIndex="0"
              >
                <div
                  className="card-image-wrapper"
                >
                  <Image
                    alt="Test Beer"
                    as="img"
                    centered={true}
                    src="https://testurl.com/beer/image.png"
                    style={
                      Object {
                        "maxHeight": 200,
                      }
                    }
                    ui={true}
                  />
                </div>
                <CardContent>
                  <CardHeader>
                    Test Beer
                  </CardHeader>
                  <CardMeta>
                    <span
                      className="date"
                    >
                      Since Just a mock beer for testing
                    </span>
                  </CardMeta>
                  <CardDescription>
                    A beer for testing
                  </CardDescription>
                </CardContent>
                <CardContent
                  extra={true}
                >
                  5 % - 40 IBU
                </CardContent>
              </Card>
            </GridColumn>,
          ],
          "columns": 4,
          "stackable": true,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "1",
            "nodeType": "function",
            "props": Object {
              "children": <Card
                centered={true}
                link={true}
                onClick={[Function]}
                onKeyPress={[Function]}
                tabIndex="0"
              >
                <div
                  className="card-image-wrapper"
                >
                  <Image
                    alt="Test Beer"
                    as="img"
                    centered={true}
                    src="https://testurl.com/beer/image.png"
                    style={
                      Object {
                        "maxHeight": 200,
                      }
                    }
                    ui={true}
                  />
                </div>
                <CardContent>
                  <CardHeader>
                    Test Beer
                  </CardHeader>
                  <CardMeta>
                    <span
                      className="date"
                    >
                      Since Just a mock beer for testing
                    </span>
                  </CardMeta>
                  <CardDescription>
                    A beer for testing
                  </CardDescription>
                </CardContent>
                <CardContent
                  extra={true}
                >
                  5 % - 40 IBU
                </CardContent>
              </Card>,
              "stretched": true,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "centered": true,
                "children": Array [
                  <div
                    className="card-image-wrapper"
                  >
                    <Image
                      alt="Test Beer"
                      as="img"
                      centered={true}
                      src="https://testurl.com/beer/image.png"
                      style={
                        Object {
                          "maxHeight": 200,
                        }
                      }
                      ui={true}
                    />
                  </div>,
                  <CardContent>
                    <CardHeader>
                      Test Beer
                    </CardHeader>
                    <CardMeta>
                      <span
                        className="date"
                      >
                        Since Just a mock beer for testing
                      </span>
                    </CardMeta>
                    <CardDescription>
                      A beer for testing
                    </CardDescription>
                  </CardContent>,
                  <CardContent
                    extra={true}
                  >
                    5 % - 40 IBU
                  </CardContent>,
                ],
                "link": true,
                "onClick": [Function],
                "onKeyPress": [Function],
                "tabIndex": "0",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <Image
                      alt="Test Beer"
                      as="img"
                      centered={true}
                      src="https://testurl.com/beer/image.png"
                      style={
                        Object {
                          "maxHeight": 200,
                        }
                      }
                      ui={true}
                    />,
                    "className": "card-image-wrapper",
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "alt": "Test Beer",
                      "as": "img",
                      "centered": true,
                      "src": "https://testurl.com/beer/image.png",
                      "style": Object {
                        "maxHeight": 200,
                      },
                      "ui": true,
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  "type": "div",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "children": Array [
                      <CardHeader>
                        Test Beer
                      </CardHeader>,
                      <CardMeta>
                        <span
                          className="date"
                        >
                          Since Just a mock beer for testing
                        </span>
                      </CardMeta>,
                      <CardDescription>
                        A beer for testing
                      </CardDescription>,
                    ],
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "function",
                      "props": Object {
                        "children": "Test Beer",
                      },
                      "ref": null,
                      "rendered": "Test Beer",
                      "type": [Function],
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "function",
                      "props": Object {
                        "children": <span
                          className="date"
                        >
                          Since Just a mock beer for testing
                        </span>,
                      },
                      "ref": null,
                      "rendered": Object {
                        "instance": null,
                        "key": undefined,
                        "nodeType": "host",
                        "props": Object {
                          "children": "Since Just a mock beer for testing",
                          "className": "date",
                        },
                        "ref": null,
                        "rendered": "Since Just a mock beer for testing",
                        "type": "span",
                      },
                      "type": [Function],
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "function",
                      "props": Object {
                        "children": "A beer for testing",
                      },
                      "ref": null,
                      "rendered": "A beer for testing",
                      "type": [Function],
                    },
                  ],
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "children": "5 % - 40 IBU",
                    "extra": true,
                  },
                  "ref": null,
                  "rendered": "5 % - 40 IBU",
                  "type": [Function],
                },
              ],
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      false,
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Grid
            columns={4}
            stackable={true}
          >
            <GridColumn
              stretched={true}
            >
              <Card
                centered={true}
                link={true}
                onClick={[Function]}
                onKeyPress={[Function]}
                tabIndex="0"
              >
                <div
                  className="card-image-wrapper"
                >
                  <Image
                    alt="Test Beer"
                    as="img"
                    centered={true}
                    src="https://testurl.com/beer/image.png"
                    style={
                      Object {
                        "maxHeight": 200,
                      }
                    }
                    ui={true}
                  />
                </div>
                <CardContent>
                  <CardHeader>
                    Test Beer
                  </CardHeader>
                  <CardMeta>
                    <span
                      className="date"
                    >
                      Since Just a mock beer for testing
                    </span>
                  </CardMeta>
                  <CardDescription>
                    A beer for testing
                  </CardDescription>
                </CardContent>
                <CardContent
                  extra={true}
                >
                  5 % - 40 IBU
                </CardContent>
              </Card>
            </GridColumn>
          </Grid>,
          false,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <GridColumn
                stretched={true}
              >
                <Card
                  centered={true}
                  link={true}
                  onClick={[Function]}
                  onKeyPress={[Function]}
                  tabIndex="0"
                >
                  <div
                    className="card-image-wrapper"
                  >
                    <Image
                      alt="Test Beer"
                      as="img"
                      centered={true}
                      src="https://testurl.com/beer/image.png"
                      style={
                        Object {
                          "maxHeight": 200,
                        }
                      }
                      ui={true}
                    />
                  </div>
                  <CardContent>
                    <CardHeader>
                      Test Beer
                    </CardHeader>
                    <CardMeta>
                      <span
                        className="date"
                      >
                        Since Just a mock beer for testing
                      </span>
                    </CardMeta>
                    <CardDescription>
                      A beer for testing
                    </CardDescription>
                  </CardContent>
                  <CardContent
                    extra={true}
                  >
                    5 % - 40 IBU
                  </CardContent>
                </Card>
              </GridColumn>,
            ],
            "columns": 4,
            "stackable": true,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "1",
              "nodeType": "function",
              "props": Object {
                "children": <Card
                  centered={true}
                  link={true}
                  onClick={[Function]}
                  onKeyPress={[Function]}
                  tabIndex="0"
                >
                  <div
                    className="card-image-wrapper"
                  >
                    <Image
                      alt="Test Beer"
                      as="img"
                      centered={true}
                      src="https://testurl.com/beer/image.png"
                      style={
                        Object {
                          "maxHeight": 200,
                        }
                      }
                      ui={true}
                    />
                  </div>
                  <CardContent>
                    <CardHeader>
                      Test Beer
                    </CardHeader>
                    <CardMeta>
                      <span
                        className="date"
                      >
                        Since Just a mock beer for testing
                      </span>
                    </CardMeta>
                    <CardDescription>
                      A beer for testing
                    </CardDescription>
                  </CardContent>
                  <CardContent
                    extra={true}
                  >
                    5 % - 40 IBU
                  </CardContent>
                </Card>,
                "stretched": true,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "centered": true,
                  "children": Array [
                    <div
                      className="card-image-wrapper"
                    >
                      <Image
                        alt="Test Beer"
                        as="img"
                        centered={true}
                        src="https://testurl.com/beer/image.png"
                        style={
                          Object {
                            "maxHeight": 200,
                          }
                        }
                        ui={true}
                      />
                    </div>,
                    <CardContent>
                      <CardHeader>
                        Test Beer
                      </CardHeader>
                      <CardMeta>
                        <span
                          className="date"
                        >
                          Since Just a mock beer for testing
                        </span>
                      </CardMeta>
                      <CardDescription>
                        A beer for testing
                      </CardDescription>
                    </CardContent>,
                    <CardContent
                      extra={true}
                    >
                      5 % - 40 IBU
                    </CardContent>,
                  ],
                  "link": true,
                  "onClick": [Function],
                  "onKeyPress": [Function],
                  "tabIndex": "0",
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": <Image
                        alt="Test Beer"
                        as="img"
                        centered={true}
                        src="https://testurl.com/beer/image.png"
                        style={
                          Object {
                            "maxHeight": 200,
                          }
                        }
                        ui={true}
                      />,
                      "className": "card-image-wrapper",
                    },
                    "ref": null,
                    "rendered": Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "function",
                      "props": Object {
                        "alt": "Test Beer",
                        "as": "img",
                        "centered": true,
                        "src": "https://testurl.com/beer/image.png",
                        "style": Object {
                          "maxHeight": 200,
                        },
                        "ui": true,
                      },
                      "ref": null,
                      "rendered": null,
                      "type": [Function],
                    },
                    "type": "div",
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "children": Array [
                        <CardHeader>
                          Test Beer
                        </CardHeader>,
                        <CardMeta>
                          <span
                            className="date"
                          >
                            Since Just a mock beer for testing
                          </span>
                        </CardMeta>,
                        <CardDescription>
                          A beer for testing
                        </CardDescription>,
                      ],
                    },
                    "ref": null,
                    "rendered": Array [
                      Object {
                        "instance": null,
                        "key": undefined,
                        "nodeType": "function",
                        "props": Object {
                          "children": "Test Beer",
                        },
                        "ref": null,
                        "rendered": "Test Beer",
                        "type": [Function],
                      },
                      Object {
                        "instance": null,
                        "key": undefined,
                        "nodeType": "function",
                        "props": Object {
                          "children": <span
                            className="date"
                          >
                            Since Just a mock beer for testing
                          </span>,
                        },
                        "ref": null,
                        "rendered": Object {
                          "instance": null,
                          "key": undefined,
                          "nodeType": "host",
                          "props": Object {
                            "children": "Since Just a mock beer for testing",
                            "className": "date",
                          },
                          "ref": null,
                          "rendered": "Since Just a mock beer for testing",
                          "type": "span",
                        },
                        "type": [Function],
                      },
                      Object {
                        "instance": null,
                        "key": undefined,
                        "nodeType": "function",
                        "props": Object {
                          "children": "A beer for testing",
                        },
                        "ref": null,
                        "rendered": "A beer for testing",
                        "type": [Function],
                      },
                    ],
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "children": "5 % - 40 IBU",
                      "extra": true,
                    },
                    "ref": null,
                    "rendered": "5 % - 40 IBU",
                    "type": [Function],
                  },
                ],
                "type": [Function],
              },
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        false,
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
